In [1]:
#TD for to dos

# for timeline analysis : prophet

## needed details 
- visit_number: number of visits on the site or the number of times they bought sth on the site
- #city_id: cityname , zipcode or sth available ?#
- price & campaign_involved: most of all what does campaign_involved denote? sometimes it seems to give a discount sometimes not 
- article_id: the id seems to be wrong ( too long )

## importing libraries

In [2]:
# Importing the necessary libraries
import re
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import glob
import datetime
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve, auc
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import confusion_matrix, f1_score, precision_recall_fscore_support
from sklearn import preprocessing
from sklearn.model_selection import cross_val_score, cross_val_predict

## setting options

In [3]:
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 400)

In [4]:
#adjusting notebook width
#from IPython.core.display import display, HTML
#display(HTML("<style>.container { width:90% !important; }</style>"))

In [5]:
#markdowntables layout

In [6]:
%%html
<style>
table {float:left}
table td, table th, table tr {text-align:left !important;}
</style>

# loading data

In [7]:
#sales dataframe
df_base = pd.read_csv("./Data/TTdata.csv")

//anaconda3/envs/nf2/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (5,20) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [8]:
#copy for working
df = df_base.copy()

In [9]:
#dataframe with citynames
#https://developers.google.com/adwords/api/docs/appendix/geotargeting
df_geotargets = pd.read_csv("./Data/geotargets.csv") 

In [10]:
df_stock = pd.read_csv("./Data/daily_TTI.DE.csv")

## first look at the data

In [11]:
df.head()

visit_id_n  unser_id_n  visit_number        date  \
0      551960      553811           1.0  20180910.0   
1      242735      254033           1.0  20180104.0   
2      680200      620544           2.0  20180612.0   
3       44920      609593          46.0  20180913.0   
4      704159       73553           2.0  20180729.0   

                article_name    article_id               ga_category  \
0  Geschenk Gutschein 10 EUR  200000000010                    Herren   
1   Geschenkgutschein 10 EUR  200000000010  Damen/Geschenkgutscheine   
2  Geschenk Gutschein 10 EUR  200000000010  Damen/Geschenkgutscheine   
3  Geschenk Gutschein 10 EUR  200000000010                    Herren   
4  Geschenk Gutschein 10 EUR  200000000010                    Herren   

       brand  price  quantity device_category device_browser device_os  \
0  (not set)   10.0       1.0         desktop        Firefox   Windows   
1  (not set)   10.0       1.0         desktop        Firefox   Windows   
2  (not set)   10.0       1.0         desktop         Chrome   Windows   
3  (not set)   10.0       1.0          mobile         Safari       iOS   
4  (not set)   10.0       1.0         desktop         Chrome   Windows   

   country        region  city_id product_line product_group product_line_r  \
0  Germany  Lower Saxony  1004478       Herren        Herren         Herren   
1  Germany  Lower Saxony  9048892       Herren        Herren         Herren   
2  Germany        Berlin  1003854       Herren        Herren         Herren   
3  Germany  Lower Saxony  1004564       Herren        Herren         Herren   
4  Germany       Bavaria  1004246       Herren        Herren         Herren   

  product_group_r campaign_involved  
0          Herren             False  
1          Herren              True  
2          Herren              True  
3          Herren              True  
4          Herren              True

In [12]:
df.tail()

visit_id_n  unser_id_n  visit_number        date  \
3016738       33526      316730           2.0  20181122.0   
3016739      520195       51192          16.0  20181201.0   
3016740       18132      393215           2.0  20181124.0   
3016741        7155      211171           NaN         NaN   
3016742        7155      211171           NaN         NaN   

                  article_name          article_id             ga_category  \
3016738  Unifarbener Ösenschal       4260478143821               (not set)   
3016739  Unifarbener Ösenschal       4260478143821  Home & Living/Vorhänge   
3016740  Unifarbener Ösenschal       4260478143821               (not set)   
3016741                    NaN          anonymized                     NaN   
3016742                    NaN  cliqz.com/tracking                     NaN   

             brand  price  quantity device_category device_browser device_os  \
3016738  TT_Casual  36.97       1.0          mobile         Chrome   Android   
3016739  TT_Casual  42.85       2.0          mobile         Safari       iOS   
3016740  TT_Casual  37.39       2.0          mobile         Safari       iOS   
3016741        NaN    NaN       NaN             NaN            NaN       NaN   
3016742        NaN    NaN       NaN             NaN            NaN       NaN   

         country                  region  city_id          product_line  \
3016738  Germany                  Berlin  1003854         Home & Living   
3016739  Germany  North Rhine-Westphalia  1004707         Home & Living   
3016740  Germany                  Berlin  1003854         Home & Living   
3016741      NaN                     NaN      NaN  Shirts unter 20 Euro   
3016742      NaN                     NaN      NaN                 Damen   

                product_group     product_line_r           product_group_r  \
3016738              Vorhänge      Home & Living      Schuhe & Accessoires   
3016739              Vorhänge      Home & Living      Schuhe & Accessoires   
3016740              Vorhänge      Home & Living      Schuhe & Accessoires   
3016741  Shirts unter 20 Euro  Shirts unter 20 €  T-Shirts & Langarmshirts   
3016742               langarm              Damen  T-Shirts & Langarmshirts   

        campaign_involved  
3016738              True  
3016739              True  
3016740              True  
3016741               NaN  
3016742               NaN

In [13]:
df.shape

(3016743, 21)

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3016743 entries, 0 to 3016742
Data columns (total 21 columns):
visit_id_n           int64
unser_id_n           int64
visit_number         float64
date                 float64
article_name         object
article_id           object
ga_category          object
brand                object
price                float64
quantity             float64
device_category      object
device_browser       object
device_os            object
country              object
region               object
city_id              object
product_line         object
product_group        object
product_line_r       object
product_group_r      object
campaign_involved    object
dtypes: float64(4), int64(2), object(15)
memory usage: 483.3+ MB


In [15]:
df.describe()

visit_id_n    unser_id_n  visit_number          date         price  \
count  3.016743e+06  3.016743e+06  3.016732e+06  3.016732e+06  3.016732e+06   
mean   4.102715e+05  3.108908e+05  1.058004e+01  2.018055e+07  2.668266e+01   
std    2.366932e+05  1.795393e+05  2.962561e+01  4.429808e+03  1.790388e+01   
min    1.000000e+00  1.000000e+00  1.000000e+00  2.017110e+07  0.000000e+00   
25%    2.053810e+05  1.551850e+05  1.000000e+00  2.018032e+07  1.445000e+01   
50%    4.107240e+05  3.107390e+05  3.000000e+00  2.018071e+07  2.420000e+01   
75%    6.152210e+05  4.659450e+05  9.000000e+00  2.018103e+07  3.361000e+01   
max    8.201950e+05  6.226180e+05  1.616000e+03  2.019023e+07  2.512600e+02   

           quantity  
count  3.016732e+06  
mean   1.018650e+00  
std    1.586700e-01  
min    1.000000e+00  
25%    1.000000e+00  
50%    1.000000e+00  
75%    1.000000e+00  
max    1.000000e+01

In [16]:
df.sample(n=30)

visit_id_n  unser_id_n  visit_number        date  \
537240       213802      329744          17.0  20171210.0   
2054870      398343      552141           6.0  20181017.0   
2027498      592864      494241          39.0  20181012.0   
2628052      425534      579522           8.0  20180826.0   
302476       352351      587536           3.0  20171222.0   
1958389      258820      341706          55.0  20190212.0   
2276280      266146      299236           3.0  20190209.0   
2051915      142306      479718          55.0  20181006.0   
525455       237849      271574           6.0  20171118.0   
2841115      583855      354274          13.0  20181003.0   
2106620       23040      126879           3.0  20181116.0   
1480063      532913      263549           9.0  20180506.0   
1236953      150944       15699           7.0  20180808.0   
334017       768084      179539           3.0  20171231.0   
1738239      563034      132917          89.0  20180816.0   
623521       323996      173670           3.0  20180109.0   
2112407      487824      572846          20.0  20190104.0   
2782478      644364      398237          74.0  20180908.0   
1294096      514741      108915           8.0  20180306.0   
757717       291227      568224           5.0  20180305.0   
1558012      222327      205580           4.0  20180527.0   
2812612       51728      158000           1.0  20181111.0   
578180        80980      529933           1.0  20171229.0   
2687932      460110        2363           2.0  20180902.0   
1890885      745287      555767           2.0  20180729.0   
1390042      506968      376453           1.0  20180528.0   
2705686      396565       37309           1.0  20180718.0   
1918700      184658       61869           6.0  20180726.0   
785296       652131      539341           3.0  20171230.0   
286120        81310      145512           3.0  20180415.0   

                               article_name     article_id  \
537240               Sweatjacke mit Sternen  4059491945808   
2054870              Langarmshirt mit Print  4060868325734   
2027498   Revolverheld: Hemd mit Stehkragen  4060868304982   
2628052             Pullover mit Rollkanten  4062105359311   
302476           Strukturiertes Damen-Kleid  4059491307163   
1958389                 Parka mit Teddyfell  4060868229582   
2276280             Schlichtes Langarmshirt  4060868882336   
2051915             Langarmshirt mit Kapuze  4060868323358   
525455       Hoodie mit Schrift-Applikation  4059491902054   
2841115                      Kariertes Hemd  4062105677354   
2106620          Offen geschnittener Blazer  4060868458302   
1480063  Loose-Fit Hose mit seitlichem Tape  4060586439379   
1236953                 gestreiftes T-Shirt  4060586196692   
334017        Herren-T-Shirt mit Logo-Print  4059491529091   
1738239            gemustertes Blusen-Shirt  4060586861934   
623521            Chino Slim Bermuda-Shorts  4059953133163   
2112407                    Gemustertes Hemd  4060868462781   
2782478                              Blazer  4062105583501   
1294096               Relaxed Tapered Jeans  4060586220717   
757717                      Troy Slim Jeans  4059953493854   
1558012                Hemd mit Brusttasche  4060586648320   
2812612         Strickpullover mit Struktur  4062105667935   
578180               Nylon-Parka mit Kapuze  4059953014257   
2687932                      Basic Cardigan  4062105499611   
1890885        Kurzarmhemd mit Brusttaschen  4060868141204   
1390042              T-Shirt mit Logo-Print  4060586327782   
2705686                 gestreiftes T-Shirt  4062105518954   
1918700            Midi-Kleid mit Stickerei  4060868159759   
785296                   Alexa Skinny Jeans  4059953530450   
286120                  T-Shirt mit Artwork  4059066815796   

                                               ga_category      brand  price  \
537240   Damen/Damenmode/Pullover & Sweatshirts/Sweatsh...   TT_Denim  18.91   
2054870                           

### feature dictionary ( my impressions )

| Feature No. | Feature Name | Description |
| :-: | :-: | :-: |
| 1 | visit_id_n | Unique ID of a single shopping cart |
| 2 | unser_id_n | Unique ID of a user (misstyped obviously) |
| 3 | visit_number | Not entirely clear; either number of visits to the shop or number of total purchases |
| 4 | date | date the purchase was made |
| 5 | article_name | Name of the article that was bought |
| 6 | article_id | ID of the article that was bought |
| 7 | ga_category | Category the article falls in ( partly the path in the webshop to the article ) |
| 8 | brand | Brand of the article |
| 9 | price | Price of the article |
| 10 | quantity | Number of the article that were bought |
| 11 | device_category | Category the user device falls in |
| 12 | device_browser | Category the user browser falls in |
| 13 | device_os | Category the user OS falls in |
| 14 | country | country location of the user |
| 15 | region | region location of the user |
| 16 | city_id | geoID of the town location of the user |
| 17 | product_line | Product line of the article |
| 18 | product_group | Product group of the article |
| 19 | product_line_r | not clear what the r stands for; might stand for retail |
| 20 | product_group_r | not clear what the r stands for; might stand for retail |
| 21 | campaign_involved | not clear what the campaign stands for |

# data cleaning

In [17]:
#rename columns
df.rename(columns={'visit_id_n':'visit_id',
                          'unser_id_n':'user_id',
                          'ga_category':'category'}, inplace=True)

## visit id
After extensive research in a later part of the project it came up, that some visit id's were assigned to multiple users. Probably because the german and austrian webshop data were just put together and not filtered. As such define visit id as the sum of visit id and user id to get a unique number.

In [18]:
df["visit_id"] = df["visit_id"].apply(lambda x: str(x)) + df["user_id"].apply(lambda x: str(x))
df = df.astype({"visit_id": "int64"})

## getting rid of nans

In [19]:
df.isna().sum()

visit_id              0
user_id               0
visit_number         11
date                 11
article_name         11
article_id            0
category             11
brand                11
price                11
quantity             11
device_category      11
device_browser       11
device_os            11
country              11
region               11
city_id              11
product_line          0
product_group         0
product_line_r        0
product_group_r       0
campaign_involved    11
dtype: int64

In [20]:
df.dropna(inplace=True) #just 11 , so drop them

In [21]:
df.isna().sum()

visit_id             0
user_id              0
visit_number         0
date                 0
article_name         0
article_id           0
category             0
brand                0
price                0
quantity             0
device_category      0
device_browser       0
device_os            0
country              0
region               0
city_id              0
product_line         0
product_group        0
product_line_r       0
product_group_r      0
campaign_involved    0
dtype: int64

In [22]:
#number of (not set)
df.isin(['(not set)']).sum(axis=0)

visit_id                  0
user_id                   0
visit_number              0
date                      0
article_name              0
article_id                0
category              41935
brand                398925
price                     0
quantity                  0
device_category           0
device_browser            0
device_os                75
country                1178
region                 7731
city_id               93307
product_line          31227
product_group         31227
product_line_r        31227
product_group_r       31227
campaign_involved         0
dtype: int64

## filtering cities and putting them into the sales dataframe

In [23]:
df_geotargets.head()

Criteria ID        Name                            Canonical Name  \
0      1000002       Kabul                   Kabul,Kabul,Afghanistan   
1      1000003      Luanda             Luanda,Luanda Province,Angola   
2      1000004  The Valley                       The Valley,Anguilla   
3      1000010   Abu Dhabi  Abu Dhabi,Abu Dhabi,United Arab Emirates   
4      1000011       Ajman          Ajman,Ajman,United Arab Emirates   

   Parent ID Country Code Target Type  Status  
0  9075393.0           AF        City  Active  
1  9070431.0           AO        City  Active  
2     2660.0           AI        City  Active  
3  9041082.0           AE        City  Active  
4  9047096.0           AE        City  Active

In [24]:
#Criteria ID as index
df_geotargets_clean = df_geotargets.set_index("Criteria ID")

In [25]:
#take just the Name column
df_geotargets_clean = df_geotargets_clean["Name"]

In [26]:
#get a dictionary from the dataframe
geodic = df_geotargets_clean.to_dict()

In [27]:
#entry for not set ids
geodic[0] = "(not set)"

In [28]:
#replace not set with a number for a moment to change to int
df["city_id"] = df["city_id"].replace({"(not set)": 0})

In [29]:
#change city_id type to int
df = df.astype({'city_id': 'int64'})

In [30]:
#map cities to dataframe with new column city
df["city"] = df["city_id"].map(geodic)

In [31]:
#nans in cities by ids that were not found
df.isna().sum()

visit_id               0
user_id                0
visit_number           0
date                   0
article_name           0
article_id             0
category               0
brand                  0
price                  0
quantity               0
device_category        0
device_browser         0
device_os              0
country                0
region                 0
city_id                0
product_line           0
product_group          0
product_line_r         0
product_group_r        0
campaign_involved      0
city                 640
dtype: int64

In [32]:
#df[df["city"].isnull()]

In [33]:
#fill the nans with (not set)
df.fillna(value={"city": "(not set)"}, inplace=True)

In [34]:
#df[df["city"].isnull()]["country"].value_counts()

In [35]:
df_geotargets["Target Type"].value_counts()

Postal Code               48391
City                      38097
Neighborhood               4631
County                     3417
Municipality               2124
Province                   1131
District                    948
Region                      909
Congressional District      441
Airport                     395
Department                  240
State                       235
University                  219
Country                     213
City Region                 183
Governorate                 121
National Park                96
Borough                      88
Prefecture                   49
Okrug                        28
Canton                       26
Autonomous Community         22
TV Region                    14
Union Territory               7
Territory                     4
Name: Target Type, dtype: int64

In [36]:
# check of the postal codes for germany ( not useful )
df_geotargets[df_geotargets["Target Type"] == "Postal Code"][df_geotargets["Country Code"]=="DE"].head()

//anaconda3/envs/nf2/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


Criteria ID   Name                        Canonical Name  Parent ID  \
66187      9041520  15306             15306,Brandenburg,Germany    20227.0   
66188      9041521  15526             15526,Brandenburg,Germany    20227.0   
66189      9041522  15230             15230,Brandenburg,Germany    20227.0   
66190      9041523  15234             15234,Brandenburg,Germany    20227.0   
66191      9041524  15236             15236,Brandenburg,Germany    20227.0   
66192      9041525  15890             15890,Brandenburg,Germany    20227.0   
66193      9041526  03172             03172,Brandenburg,Germany    20227.0   
66194      9041527  15848             15848,Brandenburg,Germany    20227.0   
66195      9041528  15907             15907,Brandenburg,Germany    20227.0   
66196      9041529  03205             03205,Brandenburg,Germany    20227.0   
66197      9041530  03222             03222,Brandenburg,Germany    20227.0   
66198      9041531  01968             01968,Brandenburg,Germany    20227.0   
66199      9041532  03238             03238,Brandenburg,Germany    20227.0   
66200      9041533  01979             01979,Brandenburg,Germany    20227.0   
66201      9041534  01987             01987,Brandenburg,Germany    20227.0   
66202      9041535  01558                  01558,Saxony,Germany    20239.0   
66203      9041536  01561                  01561,Saxony,Germany    20239.0   
66204      9041537  02977                  02977,Saxony,Germany    20239.0   
66205      9041538  01917                  01917,Saxony,Germany    20239.0   
66206      9041539  01896                  01896,Saxony,Germany    20239.0   
66207      9041540  02625                  02625,Saxony,Germany    20239.0   
66208      9041541  03130             03130,Brandenburg,Germany    20227.0   
66209      9041542  03046             03046,Brandenburg,Germany    20227.0   
66210      9041543  03044             03044,Brandenburg,Germany    20227.0   
66211      9041544  03149             03149,Brandenburg,Germany    20227.0   
66212      9041545  02943                  02943,Saxony,Germany    20239.0   
66213      9041546  02826                  02826,Saxony,Germany    20239.0   
66214      9041547  02827                  02827,Saxony,Germany    20239.0   
66215      9041548  02708                  02708,Saxony,Germany    20239.0   
66216      9041549  02906                  02906,Saxony,Germany    20239.0   
66217      9041550  02763                  02763,Saxony,Germany    20239.0   
66218      9041551  02730                  02730,Saxony,Germany    20239.0   
66219      9041552  01877                  01877,Saxony,Germany    20239.0   
66220      9041553  01844                  01844,Saxony,Germany    20239.0   
66221      9041554  01328                  01328,Saxony,Germany    20239.0   
66222      9041555  01809                  01809,Saxony,Germany    20239.0   
66223      9041556  01796                  01796,Saxony,Germany    20239.0   
66224      9041557  01744                  01744,Saxony,Germany    20239.0   
66225      9041558  01728                  01728,Saxony,Germany    20239.0   
66226      9041559  01705                  01705,Saxony,Germany    20239.0   
66227      9041560  01187                  01187,Saxony,Germany    20239.0   
66228      9041561  01159                  01159,Saxony,Germany    20239.0   
66229      9041562  01217                  01217,Saxony,Germany    20239.0   
66230      9041563  01069                  01069,Saxony,Germany    20239.0   
66231      9041564  01219                  01219,Saxony,Germany    20239.0   
66232      9041565  01277                  01277,Saxony,Germany    20239.0   
66233      9041566  01307                  01307,Saxony,Germany    20239.0   
66234      9041567  01309                  01309,Saxony,Germany    20239.0   
66235      9041568  01099                  01099,Saxony,Germany    20239.0   
66236      9041569  01454                  01454,Saxony,Germany    20239.0   
66237      9041570  01458    

In [37]:
#Brand seems inherintly useless 
df.brand.unique()

array(['(not set)', 'TT_Casual', 'TT_Denim'], dtype=object)

### price seems wrong for the following, as the original price is around 5-8, so discounts are in the price already

In [38]:
df.iloc[658646]

visit_id                                                  424670120783
user_id                                                         120783
visit_number                                                         4
date                                                       2.01807e+07
article_name                                  Bandana in Melange-Optik
article_id                                               4059953239858
category             Kinder/Kindermode/Baby Mädchen 0 - 17 Monate/A...
brand                                                        TT_Casual
price                                                             0.38
quantity                                                             1
device_category                                                desktop
device_browser                                                 Firefox
device_os                                                      Windows
country                                                        Germany
region

## change datatypes

In [39]:
#change visit_number and quantity to int as float does not make sense
df = df.astype({'visit_number': 'int64', "quantity": "int64", "article_id": "int64"})
#device stuff and brand to category
df = df.astype({'device_category': 'category', "device_browser": "category", "device_os": "category", "brand": "category"})
#campaign_involved to bool
df.campaign_involved = df.campaign_involved.astype("bool")

In [40]:
#change objects to string for starters
#df[df.select_dtypes('object').columns] = df[df.select_dtypes('object').columns].astype(np.str) 
#does not work

In [41]:
#date to date format
df["date"]= pd.to_datetime(df["date"],format='%Y%m%d')

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3016732 entries, 0 to 3016740
Data columns (total 22 columns):
visit_id             int64
user_id              int64
visit_number         int64
date                 datetime64[ns]
article_name         object
article_id           int64
category             object
brand                category
price                float64
quantity             int64
device_category      category
device_browser       category
device_os            category
country              object
region               object
city_id              int64
product_line         object
product_group        object
product_line_r       object
product_group_r      object
campaign_involved    bool
city                 object
dtypes: bool(1), category(4), datetime64[ns](1), float64(1), int64(6), object(9)
memory usage: 428.7+ MB


## list of years to look at the span of time we have

In [43]:
year_list = pd.DatetimeIndex(df['date']).year  

In [44]:
year_list.unique()

Int64Index([2018, 2017, 2019], dtype='int64', name='date')

In [45]:
df.date.max()

Timestamp('2019-02-28 00:00:00')

In [46]:
df.date.min()

Timestamp('2017-11-01 00:00:00')

We have data from 2017-11-01 till 2019-02-28

## check for missing dates 

In [47]:
from datetime import date, timedelta
date_set = set(df.date.min() + timedelta(x) for x in range((df.date.max() - df.date.min()).days))
missing = sorted(date_set - set(df.date))
missing

[]

As we can see there are no missing days in the date Data.

## add days
monday 0 , ... , sunday 6

In [48]:
df["day_of_week"] = pd.to_datetime(df["date"]).dt.dayofweek

## article names and id

In [49]:
df.article_name.value_counts()

Alexa Slim Jeans                                  62458
Josh Regular Slim Jeans                           41092
T-Shirt mit Schrift-Print                         38895
T-Shirt mit Print                                 29257
T-Shirt mit Logo-Print                            28231
Marvin Straight Jeans                             27405
Alexa Straight Jeans                              22816
Damen-Basic-Pullover mit V-Ausschnitt             19512
Jim Slim Bermuda Shorts                           18002
Basic Herren-Pullover mit V-Ausschnitt            17812
Travis Regular Chino                              16859
Basic Herren-Strick-Pullover                      14865
gemustertes Blusen-Shirt                          14560
Gemusterte Bluse                                  14144
gemustertes T-Shirt                               13267
Jona Extra Skinny Jeans                           13179
Gestreifter Damen-Pullover                        13091
Damen-Straight-Jeans – Alexa                    

In [50]:
df.article_id.value_counts()

4059066698719    1728
4059066698511    1284
4002702235090    1215
4059066698726    1176
4059066698733    1054
4059953383155     998
4059953383148     908
4059953383162     887
4059953383179     861
4059491897060     819
4060586142743     799
4060586158591     791
4058114237092     789
4002702235083     783
4059953383049     781
4052691223851     781
4054985602212     765
4052691223844     751
4002702235076     748
4060586158607     743
4060586142781     725
4060586230112     718
4059066698672     709
4059491897077     697
4062105707020     682
4059953383186     678
4059491843289     677
4059491843302     675
4062105681436     668
4056762815792     659
4059491784308     658
4059491520005     653
4058114237108     651
4062105692074     648
4054985761858     646
4059491784315     645
4060586158614     636
4054985761865     632
4059491843296     631
4057655756987     628
4060586270033     627
4059491776532     625
4060586262861     621
4059953863794     617
4059953383193     613
4056762815

## change the names of the following to match

In [51]:
d = {'Geschenk Gutschein 10 EUR':'Geschenkgutschein 10 EUR','Geschenk Gutschein 20 EUR':'Geschenkgutschein 20 EUR','Geschenk Gutschein 50 EUR':'Geschenkgutschein 50 EUR'}
df = df.replace(d)

## duplicated rows

In [52]:
df.duplicated().sum()

632

In [53]:
duplicateRowsDF = df[df.duplicated(keep=False)]

In [54]:
duplicateRowsDF.head()

visit_id  user_id  visit_number       date  \
118    22892413921   413921            45 2018-07-11   
119    22892413921   413921            45 2018-07-11   
675   177945413228   413228             1 2019-01-01   
676   177945413228   413228             1 2019-01-01   
1010  601457557264   557264             1 2018-12-25   

                      article_name     article_id  \
118       Geschenkgutschein 20 EUR   200000000020   
119       Geschenkgutschein 20 EUR   200000000020   
675   Sneaker-Socken im Doppelpack  4002702123007   
676   Sneaker-Socken im Doppelpack  4002702123007   
1010  Sneaker-Socken im Doppelpack  4002702123014   

                                               category      brand  price  \
118                                              Herren  (not set)  20.00   
119                                              Herren  (not set)  20.00   
675   Kinder/Kindermode/Mädchen 1,5 - 7  Jahre/Schuh...  TT_Casual   2.31   
676   Kinder/Kindermode/Mädchen 1,5 - 7  Jahre/Schuh...  TT_Casual   2.31   
1010  Kinder/Kindermode/Mädchen 1,5 - 7  Jahre/Schuh...  TT_Casual   3.18   

      quantity device_category device_browser device_os  country  \
118          1         desktop         Chrome   Windows  Austria   
119          1         desktop         Chrome   Windows  Austria   
675          1          mobile         Chrome   Android  Germany   
676          1          mobile         Chrome   Android  Germany   
1010        10          mobile         Chrome       iOS  Germany   

                 region  city_id product_line         product_group  \
118       Upper Austria  1000827       Herren                Herren   
119       Upper Austria  1000827       Herren                Herren   
675              Berlin  1003854       Kinder  Schuhe & Accessoires   
676              Berlin  1003854       Kinder  Schuhe & Accessoires   
1010  Baden-Wurttemberg  1004022       Kinder  Schuhe & Accessoires   

     product_line_r       product_group_r  campaign_involved     city  \
118          Herren                Herren              False     Linz   
119          Herren                Herren              False     Linz   
675          Kinder  Schuhe & Accessoires               True   Berlin   
676          Kinder  Schuhe & Accessoires               True   Berlin   
1010         Kinder  Schuhe & Accessoires              False  Rastatt   

      day_of_week  
118             2  
119             2  
675             1  
676             1  
1010            1

Note that it is not sure whether these rows are truly duplicates or just multiple purchases where the quantity wasn't set properly or the articles have the same name and price and for example different colors which this data does not tell us.

In [55]:
#df.drop_duplicates(keep="first", inplace=True)
#df.reset_index()

checking number of visits to number of entries (45 for this user)

In [56]:
df.user_id[413921]

230677

In [57]:
df[df['user_id'] == 413921].shape

(27, 23)

## duplicates in visit_id

In [58]:
df.visit_id.max()

820195135969

In [59]:
df.visit_id.duplicated().sum()

2181687

In [60]:
df_visits = df[df.duplicated(subset="visit_id") == True]

In [61]:
df_uniquevisits = df.drop_duplicates(subset="visit_id")

In [62]:
df_visits.shape

(2181687, 23)

In [63]:
df_uniquevisits.shape

(835045, 23)

In [64]:
df_visits.sort_values(by=['visit_id']).head()

visit_id  user_id  visit_number       date  \
2972494    234440       40             1 2018-11-18   
2817928    234440       40             1 2018-11-18   
2538377    234440       40             1 2018-11-18   
2375327    276103    76103            22 2019-02-21   
2647054    290200      200            43 2018-11-03   

                         article_name     article_id  \
2972494              Schlichter Snood  4062105772851   
2817928              Hoodie mit Print  4062105670652   
2538377  Basic Herren-Strick-Pullover  4062105188904   
2375327          Culver Skinny Jeans   4061945274082   
2647054                    Steppjacke  4062105383903   

                               category      brand  price  quantity  \
2972494                  Hoodies/Herren   TT_Denim  17.99         1   
2817928                  Hoodies/Herren  TT_Casual  15.00         1   
2538377                          Herren  TT_Casual  14.99         1   
2375327  Herren/Herrenmode/Jeans/Skinny   TT_Denim  42.01         1   
2647054                     Steppjacken  TT_Casual  31.51         1   

        device_category device_browser device_os  country         region  \
2972494          mobile         Chrome   Android  Austria  Lower Austria   
2817928          mobile         Chrome   Android  Austria  Lower Austria   
2538377          mobile         Chrome   Android  Austria  Lower Austria   
2375327          mobile         Safari       iOS  Germany         Berlin   
2647054          mobile         Chrome   Android  Germany    Brandenburg   

         city_id    product_line   product_group  product_line_r  \
2972494  1000750          Herren     Sweatshirts          Herren   
2817928  1000750         Hoodies          Herren          Herren   
2538377  1000750          Herren        Pullover          Herren   
2375327  1003854          Herren          Skinny          Herren   
2647054  1003866  Herbst Outfits  Herbst Outfits  Herbst Outfits   

                product_group_r  campaign_involved                   city  \
2972494  Pullover & Sweatshirts               True                   Melk   
2817928                  Herren               True                   Melk   
2538377  Pullover & Sweatshirts               True                   Melk   
2375327                   Jeans               True                 Berlin   
2647054          Herbst Outfits               True  Frankfurt an der Oder   

         day_of_week  
2972494            6  
2817928            6  
2538377            6  
2375327            3  
2647054            5

In [65]:
df_uniquevisits.sort_values(by=["visit_id"]).head()

visit_id  user_id  visit_number       date  \
320852     234440       40             1 2018-11-18   
2079241    276103    76103            22 2019-02-21   
2645748    290200      200            43 2018-11-03   
2830805    336120       20             2 2018-10-25   
321178     366022     6022             8 2017-12-30   

                         article_name     article_id  \
320852   Basic Herren-Strick-Pullover  4059491520210   
2079241               Henley T-Shirt   4060868345305   
2645748                    Steppjacke  4062105383569   
2830805       Nela Extra Skinny Jeans  4062105674247   
321178   Basic Herren-Strick-Pullover  4059491520241   

                                                  category      brand  price  \
320852                                              Herren  TT_Casual  14.99   
2079241  Herren/Herrenmode/T-Shirts & Langarmshirts/T-S...  TT_Casual  10.91   
2645748                                  Steppjacken/Damen  TT_Casual  31.50   
2830805                       Damen/Damenmode/Jeans/Skinny   TT_Denim  33.61   
321178                                    Feeling Blue Men  TT_Casual  16.80   

         quantity device_category device_browser  device_os  country  \
320852          1          mobile         Chrome    Android  Austria   
2079241         1          mobile         Safari        iOS  Germany   
2645748         1          mobile         Chrome    Android  Germany   
2830805         1          mobile         Chrome    Android  Germany   
321178          1         desktop         Safari  Macintosh  Germany   

                region  city_id      product_line     product_group  \
320852   Lower Austria  1000750            Herren          Pullover   
2079241         Berlin  1003854            Herren          T-Shirts   
2645748    Brandenburg  1003866             Damen            Jacken   
2830805       Saarland  9048444             Damen            Skinny   
321178          Berlin  1003854  Feeling Blue Men  Feeling Blue Men   

           product_line_r           product_group_r  campaign_involved  \
320852             Herren    Pullover & Sweatshirts               True   
2079241            Herren  T-Shirts & Langarmshirts               True   
2645748             Damen   Jacken, Mäntel & Westen               True   
2830805             Damen                     Jeans               True   
321178   Feeling Blue Men          Feeling Blue Men               True   

                          city  day_of_week  
320852                    Melk            6  
2079241                 Berlin            3  
2645748  Frankfurt an der Oder            5  
2830805                Homburg            3  
321178                  Berlin            5

Duplicates in the visit_id surely denote multiple articles that were bought at the same visit, thus dropping any will not make sense. Also there were 820194 unique visit_id's, hence we know that the data consists of this number of shopping cart purchases.

## unique devices 
lets check the entries and number of occurance of the elements of the device features

In [68]:
df.device_browser.value_counts()

Chrome                      994731
Safari                      793300
Firefox                     589700
Samsung Internet            256273
Edge                        155931
Internet Explorer            88370
Android Webview              63918
Opera                        29834
Safari (in-app)              28517
Amazon Silk                  14662
YaBrowser                      354
SeaMonkey                      337
Puffin                         235
Maxthon                        217
Android Browser                 94
Mozilla Compatible Agent        85
Iron                            58
UC Browser                      43
Playstation 4                   26
BlackBerry                      14
MRCHROME                        12
Seznam                          12
Coc Coc                          4
Opera 9.4                        3
Nintendo Browser                 2
Name: device_browser, dtype: int64

In [69]:
df.device_category.value_counts()

desktop    1371610
mobile     1091610
tablet      553512
Name: device_category, dtype: int64

In [70]:
df.device_os.value_counts()

Windows          1204420
Android           923303
iOS               718614
Macintosh         156569
Linux              11633
Chrome OS           1778
BlackBerry           222
(not set)             75
Windows Phone         70
Playstation 4         26
Tizen                 19
Nintendo WiiU          2
FreeBSD                1
Name: device_os, dtype: int64

Nothing too surprising so far. Later we can check if there were differences in the purchases between the different features.

## campaign_involved:

In [71]:
camp_true = df[df.campaign_involved == True]

In [72]:
len(camp_true)

2278279

In [73]:
camp_true.sample(n=20).head()

visit_id  user_id  visit_number       date  \
2145111   67073857439    57439             1 2019-01-08   
203547   156451590761   590761            24 2018-03-25   
2437131  466646298926   298926            15 2018-07-18   
686617   332814312259   312259            69 2018-01-19   
2158059  707616520581   520581             5 2018-10-08   

                         article_name     article_id  \
2145111        Schlichtes Blusenshirt  4060868494959   
203547   Sneaker mit Metallic-Details  4058219388705   
2437131       Gemusterte Kurzarmbluse  4062105065571   
686617      Pullover mit V-Ausschnitt  4059953296301   
2158059                Karierte Bluse  4060868502746   

                                                category      brand  price  \
2145111                   Damen/Damenmode/Blusen/Langarm  TT_Casual  22.05   
203547            Damen/Damenschuhe & Accessoires/Schuhe  (not set)  32.32   
2437131                   Damen/Damenmode/Blusen/kurzarm  TT_Casual  16.80   
686617   Damen/Damenmode/Pullover & Sweatshirts/Pullover  TT_Casual  26.88   
2158059                   Damen/Damenmode/Blusen/Langarm  TT_Casual  20.16   

         quantity device_category device_browser device_os  country  \
2145111         1         desktop         Chrome   Windows  Germany   
203547          1         desktop        Firefox   Windows  Germany   
2437131         1          tablet         Chrome   Android  Germany   
686617          1         desktop         Chrome     Linux  Germany   
2158059         1          tablet         Safari       iOS  Germany   

                         region  city_id product_line product_group  \
2145111  North Rhine-Westphalia  1004636        Damen         Damen   
203547                    Hesse  1004422        Damen        Schuhe   
2437131       Baden-Wurttemberg  1004066        Damen       kurzarm   
686617             Lower Saxony  1004536        Damen   Neu im Sale   
2158059                 Bavaria  1004314        Damen       Langarm   

        product_line_r           product_group_r  campaign_involved  \
2145111          Damen                     Damen               True   
203547           Damen      Schuhe & Accessoires               True   
2437131          Damen  T-Shirts & Langarmshirts               True   
686617           Damen               Neu im Sale               True   
2158059          Damen  T-Shirts & Langarmshirts               True   

                           city  day_of_week  
2145111            Grevenbroich            1  
203547              Russelsheim            6  
2437131  Villingen-Schwenningen            2  
686617                Oldenburg            4  
2158059  Vilshofen an der Donau            0

In [74]:
camp_false = df[df.campaign_involved == False]

In [75]:
len(camp_false)

738453

In [76]:
camp_false[camp_false["article_id"]==4011694579266].head()

visit_id  user_id  visit_number       date           article_name  \
35271  505889355464   355464             5 2017-11-13  Hip-Pants im 3er-Pack   
35274  383063313091   313091             2 2018-03-24  Hip-Pants im 3er-Pack   
35277  201946434776   434776             6 2018-07-29  Hip-Pants im 3er-Pack   
35281  333769606504   606504             1 2018-09-17  Hip-Pants im 3er-Pack   
35295   14858564131   564131             1 2019-02-04  Hip-Pants im 3er-Pack   

          article_id                               category      brand  price  \
35271  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  15.12   
35274  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  18.90   
35277  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  25.20   
35281  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  20.16   
35295  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  16.80   

       quantity device_category device_browser  device_os  country  \
35271         1          mobile        Firefox    Android  Germany   
35274         1          tablet         Safari        iOS  Germany   
35277         1         desktop        Firefox    Windows  Germany   
35281         1         desktop         Chrome  Macintosh  Germany   
35295         1         desktop        Firefox  Macintosh  Germany   

                   region  city_id product_line        product_group  \
35271              Berlin  1003854       Herren  Underwear & Pyjamas   
35274  Schleswig-Holstein  9048708       Herren  Underwear & Pyjamas   
35277  Schleswig-Holstein  1004866       Herren  Underwear & Pyjamas   
35281              Berlin  1003854       Herren  Underwear & Pyjamas   
35295        Lower Saxony  1004536       Herren  Underwear & Pyjamas   

      product_line_r      product_group_r  campaign_involved  \
35271         Herren  Underwear & Pyjamas              False   
35274         Herren  Underwear & Pyjamas              False   
35277         Herren  Underwear & Pyjamas              False   
35281         Herren  Underwear & Pyjamas              False   
35295         Herren  Underwear & Pyjamas              False   

                     city  day_of_week  
35271              Berlin            0  
35274  Sankt Peter-Ording            5  
35277                Kiel            6  
35281              Berlin            0  
35295           Oldenburg            0

In [77]:
camp_true[camp_true["article_id"]==4011694579266].head()

visit_id  user_id  visit_number       date           article_name  \
35270  723099348206   348206             6 2017-11-13  Hip-Pants im 3er-Pack   
35272  539377527591   527591             6 2018-08-29  Hip-Pants im 3er-Pack   
35273   54340111696   111696             2 2018-09-08  Hip-Pants im 3er-Pack   
35275  492320350403   350403             6 2018-10-26  Hip-Pants im 3er-Pack   
35276   31699418198   418198             1 2018-09-13  Hip-Pants im 3er-Pack   

          article_id                               category      brand  price  \
35270  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  21.42   
35272  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  25.21   
35273  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  18.90   
35275  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  14.70   
35276  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  18.90   

       quantity device_category    device_browser  device_os  country  \
35270         1          mobile  Samsung Internet    Android  Germany   
35272         1          mobile  Samsung Internet    Android  Germany   
35273         1         desktop            Chrome  Chrome OS  Germany   
35275         2         desktop            Chrome    Windows  Germany   
35276         1          mobile            Chrome    Android  Germany   

                       region  city_id product_line        product_group  \
35270                  Berlin  1003854       Herren  Underwear & Pyjamas   
35272                 Bavaria  1004269       Herren  Underwear & Pyjamas   
35273            Lower Saxony  9048874       Herren  Underwear & Pyjamas   
35275  North Rhine-Westphalia  1004761       Herren  Underwear & Pyjamas   
35276               Thuringia  1004999       Herren  Underwear & Pyjamas   

      product_line_r      product_group_r  campaign_involved            city  \
35270         Herren  Underwear & Pyjamas               True          Berlin   
35272         Herren  Underwear & Pyjamas               True       Rosenheim   
35273         Herren  Underwear & Pyjamas               True        Wiesmoor   
35275         Herren  Underwear & Pyjamas               True  Wermelskirchen   
35276         Herren  Underwear & Pyjamas               True            Suhl   

       day_of_week  
35270            0  
35272            2  
35273            5  
35275            4  
35276            3

we see, that campaign_involved can not mean discounts

## nr. of occurence of countries

In [78]:
df.country.value_counts()

Germany                  2756836
Austria                   196950
Netherlands                22547
Switzerland                 5425
Belgium                     2434
Czechia                     2293
Poland                      2174
United Kingdom              2021
Italy                       1956
United States               1850
Spain                       1827
Hungary                     1766
Slovenia                    1457
France                      1425
Slovakia                    1366
Romania                     1331
(not set)                   1178
Estonia                     1038
Denmark                      969
Croatia                      900
Luxembourg                   852
Ukraine                      842
Lithuania                    804
Latvia                       710
Sweden                       514
Greece                       507
Turkey                       466
Bulgaria                     348
United Arab Emirates         258
Portugal                     248
Ireland   

## columns for article type

In [79]:
#put all features describing the article type together for easier use
df["genre_column_placeholder"] = df["article_name"] + "/" + df["category"] + "/" + df["product_line"] + "/" + df["product_line_r"] + "/" + df["product_group"] + "/" + df["product_group_r"]

In [80]:
df.genre_column_placeholder = df.genre_column_placeholder.astype("str")

In [81]:
#new column for articles of women genre
def check_women(item):
    return "" if re.search(r"([Dd]{1}ame[ns]{1})|([Ww]{1}om[ea]{1}n)", item) == None else "Women"

df['women_article'] = df['genre_column_placeholder'].apply(check_women)


In [82]:
#new column for articles of men genre
def check_men(item):
    return "" if re.search(r"([Hh]{1}e[r]{1,2}en)|(Men\b)|(\b[Mm]en\b)", item) == None else "Men"

df['men_article'] = df['genre_column_placeholder'].apply(check_men)


In [83]:
#new column for articles of kids/teenager/babies genre
def check_kids(item):
    return "" if re.search(r"([Kk]{1}inder)|(\b[Kk]{1}id[s]{0,1})|([Jj]{1}ungen)|([Mm]{1}ädchen)|([Tt]{1}een[s]{0,1})|([Bb]{1}aby)|([Bb]{1}abies)", item) == None else "Kids"

df['kids_article'] = df['genre_column_placeholder'].apply(check_kids)


In [84]:
#new column for articles of home&living genre
def check_hl(item):
    return "" if re.search(r"([Hh]{1}ome)|([Ll]iving)", item) == None else "Home&living"

df['homeandliving_article'] = df['genre_column_placeholder'].apply(check_hl)


In [85]:
df = df.assign(category_slim = df.women_article + df.men_article + df.kids_article + df.homeandliving_article)

In [86]:
def putslash(item): 
  
    # regex [A-Z][a-z]* means any string starting  
    # with capital character followed by many  
    # lowercase letters  
    words = re.findall('[A-Z][a-z]*', item) 
  
    # Change first letter of each word into lower 
    # case 
    result = [] 
    for word in words: 
        word = chr( ord (word[0]) + 32) + word[1:] 
        result.append(word) 
    return '/'.join(result)

In [87]:
df["category_slim"] = df["category_slim"].apply(putslash)

In [88]:
df["category_slim"] = df["category_slim"].replace("", "(not set)")

In [90]:
df.women_article = df.women_article.replace({"": 0, "Women": 1})
df.men_article = df.men_article.replace({"": 0, "Men": 1})
df.kids_article = df.kids_article.replace({"": 0, "Kids": 1})
df.homeandliving_article = df.homeandliving_article.replace({"": 0, "Home&living": 1})

In [120]:
article_diff = df[(df.women_article + df.men_article + df.kids_article + df.homeandliving_article) == 0]

In [121]:
article_diff.product_group_r.value_counts()

Schuhe & Accessoires        7878
(not set)                   7284
Herbst Outfits              6384
T-Shirts & Langarmshirts    1042
Neu im Sale                  427
Festival Styles              326
SAFARI STYLES                212
Summer Sale Highlights       121
Bold Summer Colors            83
Black & White                 46
Alles zum halben Preis        24
T-Shirts unter 20 €            9
Partyoutfits                   3
Revolverheld Kollektion        2
Mexico Styles                  2
Name: product_group_r, dtype: int64

In [122]:
article_diff.category.value_counts()

(not set)                                  7000
Herbst Outfits                             4431
Athleisure                                 3881
Schuhe                                     1805
Winterpullover                              818
Business-Kleider                            462
Neu im Sale                                 420
Baumwollkleidung                            403
Gürtel                                      380
Shirts unter 20 Euro                        377
Baumwollhosen                               364
Taschen/medium                              324
Kurzblazer                                  259
Slim                                        232
SAFARI STYLES                               190
Sonnenbrillen                               185
medium                                      152
Festivalstyles                              146
Taschen                                     138
T-Shirts unter 20 Euro                       91
Summer Sale Highlights                  

In [123]:
len(article_diff)

23843

So we are left with only 23843 articles without any of the new article groups.

In [94]:
#number of articles that got multiple article groups
mult_article = df[df.women_article + df.men_article + df.kids_article + df.homeandliving_article > 1]

In [95]:
mult_article.sample(n=20)

visit_id  user_id  visit_number       date  \
1361734  311721312008   312008             4 2018-03-15   
257887    67382209390   209390             1 2019-01-09   
2113977  634027560197   560197             1 2018-12-28   
2681721   33135984437    84437             3 2018-12-23   
1972972  270592569585   569585             3 2018-09-02   
2768834   96238216622   216622             1 2018-10-10   
1225858  401271291386   291386             1 2018-12-24   
2018498  229197164115   164115             7 2018-10-27   
2677410  760650154549   154549             4 2018-09-22   
2677065  164887272909   272909             4 2018-09-11   
90682      7390714544     4544             2 2018-05-18   
2113769  391237363694   363694            22 2019-01-24   
2742684  356064603139   603139             9 2018-10-05   
987944   388597392917   392917             3 2018-04-11   
2531279  503055104974   104974            80 2018-09-26   
1227007  789250444250   444250             4 2018-05-25   
2584543  293111471824   471824             4 2018-08-02   
2259368  323141168743   168743            14 2018-11-20   
2338773  671893244244   244244            18 2018-11-25   
2934898  646002182960   182960             1 2018-12-12   

                                    article_name     article_id  \
1361734                  Lässige Damen-Stoffhose  4060586270026   
257887                      Travis Regular Chino  4059066344227   
2113977                  Nela Extra Skinny Jeans  4060868464655   
2681721                          Alexa Slim Hose  4062105491066   
1972972                               Steppjacke  4060868238713   
2768834                             Winter Parka  4062105566351   
1225858                  gemustertes Kurzarmhemd  4060586194452   
2018498                    Kleid mit Taillenbund  4060868300090   
2677410                          Kunstlederjacke  4062105464695   
2677065                          Kunstlederjacke  4062105464602   
90682              Herren-T-Shirt mit Logo-Print  4053078743238   
2113769                  Nela Extra Skinny Jeans  4060868464501   
2742684                  Gefütterter Steppmantel  4062105536637   
987944                     Jacke im Material-Mix  4060463021314   
2531279             Basic Herren-Strick-Pullover  4062105174990   
1227007                  gemustertes Kurzarmhemd  4060586194537   
2584543                        Mia Slim 7/8 Hose  4062105311838   
2259368  Strickpullover mit Blockstreifen-Muster  4060868837770   
2338773                Bluse mit seitlichem Tape  4061945167711   
2934898                              Skaterkleid  4062105703527   

                                                  category      brand  price  \
1361734                              Damen/Damenmode/Hosen   TT_Denim  26.88   
257887                             Herren/Herrenmode/Hosen  TT_Casual  28.35   
2113977                       Damen/Damenmode/Jeans/Skinny   TT_Denim  45.16   
2681721                              Damen/Damenmode/Hosen  TT_Casual  33.61   
1972972     Damen/Damenmode/Jacken, Mäntel & Westen/Jacken   TT_Denim  58.82   
2768834                                Winterparkas/Herren  TT_Casual  67.22   
1225858                   Herren/Herrenmode/Hemden/Kurzarm  TT_Casual  15.96   
2018498                                Business-Mode/Damen  TT_Casual  28.00   
2677410                             Uebergangsjacken/Damen  TT_Casual  83.32   
2677065                                       Sommerjacken  TT_Casual  67.22   
90682                                  Herren/Print Shirts  TT_Casual   7.52   
2113769                       Damen/Damenmode/Jeans/Skinny   TT_Denim  40.60   
2742684                                              Damen  TT_Casual  85.71   
987944   Kinder/Kindermode/Teens Jungen 8 - 16 Jahre/Ja...  (not set)  50.41   
2531279  Herren/Herrenmode/Pullover & Sweatshirts/Pullover  TT_Casual  20.16   
1227007                   Herren/Herrenmode/Hemden/kurzarm  TT_Casual  18.90   
25845

In [96]:
len(mult_article)

65808

Having 65808 articles with multiple article groups. Most of them do make at least some sense, like a jacket for a teenager also having the product_line Men.

## total price per article

In [97]:
df["total_price"] = df["price"] * df["quantity"]

## ordering columns in a proper way

In [99]:
#code for checking the index of a certain column
df.columns.get_loc("category_slim")

28

In [100]:
#check name of certain index
df.columns[20]

'campaign_involved'

In [101]:
#reordering the columns
cols = df.columns.tolist()
cols = cols[0:4] + [cols[22]] + cols[4:6] + cols[8:10] + [cols[29]] \
        + cols[13:15] + [cols[21]] + [cols[15]] + cols[24:28] + [cols[20]] \
        + [cols[6]] + [cols[28]] + cols[7:8] + cols[16:20] + [cols[23]] + cols[10:13]
print("Check if any column was skipped when reordering (True: all columns fetched):", len(cols) == len(df.columns))
df = df.reindex(columns = cols)

Check if any column was skipped when reordering (True: all columns fetched): True


## adding value and number of items in the shopping cart of a visit

In [102]:
shopcart = df.groupby(["visit_id"]).sum()

In [103]:
#take just the Name column
df_shopcartvalue_clean = shopcart["total_price"]
df_shopcartnumber_clean = shopcart["quantity"]

In [104]:
#get a dictionary from the dataframe
cartdicval = df_shopcartvalue_clean.to_dict()
cartdicnum = df_shopcartnumber_clean.to_dict()
cartdiffart = df.visit_id.value_counts().to_dict()

In [105]:
df["shop_cart_value"] = df["visit_id"].map(cartdicval)
df["shop_cart_item_count"] = df["visit_id"].map(cartdicnum)
df["shop_cart_unique_articles"] = df["visit_id"].map(cartdiffart)

In [107]:
#reordering the columns
cols = df.columns.tolist()
cols = cols[0:10] + cols[30:33] + cols[10:30]
print("Check if any column was skipped when reordering (True: all columns fetched):", len(cols) == len(df.columns))
df = df.reindex(columns = cols)

Check if any column was skipped when reordering (True: all columns fetched): True


## add stockprice per day to the dataframe

In [109]:
df_stock["timestamp"]= pd.to_datetime(df_stock["timestamp"],format='%Y-%m-%d')

In [110]:
#fill missing days ( weekends )
df_stock = df_stock.resample('D',on="timestamp").mean()

In [111]:
#fill the nans of the missings above with the average of their direct set neighbours
df_stock = (df_stock.ffill() + df_stock.bfill())/2

In [112]:
df_stock.head()

open      high      low    close     volume
timestamp                                                 
2010-03-26  11.9637  12.04200  11.4576  11.8257  2767746.0
2010-03-27  11.8740  11.97755  11.5404  11.8349  1489778.5
2010-03-28  11.8740  11.97755  11.5404  11.8349  1489778.5
2010-03-29  11.7843  11.91310  11.6232  11.8441   211811.0
2010-03-30  11.8073  11.91770  11.6600  11.8717    80096.0

In [113]:
#take just the close column
df_stock_clean = df_stock["close"]

In [114]:
#get a dictionary from the dataframe
stockdic = df_stock_clean.to_dict()

In [115]:
#map stock to dataframe with new column stock_value
df["stock_value"] = df["date"].map(stockdic)

In [116]:
#check for nans that were not set 
df.isna().sum()

visit_id                     0
user_id                      0
visit_number                 0
date                         0
day_of_week                  0
article_name                 0
article_id                   0
price                        0
quantity                     0
total_price                  0
shop_cart_value              0
shop_cart_item_count         0
shop_cart_unique_articles    0
country                      0
region                       0
city                         0
city_id                      0
women_article                0
men_article                  0
kids_article                 0
homeandliving_article        0
campaign_involved            0
category                     0
category_slim                0
brand                        0
product_line                 0
product_group                0
product_line_r               0
product_group_r              0
genre_column_placeholder     0
device_category              0
device_browser               0
device_o

In [117]:
#easy dataframe to check the stockprice development 
df.drop_duplicates("date").sort_values("date").head(20)

visit_id  user_id  visit_number       date  day_of_week  \
2325   14930374362   374362             3 2017-11-01            2   
5836  645245525600   525600             2 2017-11-02            3   
2849  448437326618   326618             1 2017-11-03            4   
2512  813967575820   575820            11 2017-11-04            5   
455   754081156504   156504             1 2017-11-05            6   
357   817688334365   334365             4 2017-11-06            0   
4025  403206555145   555145             3 2017-11-07            1   
2574   81446267456   267456             2 2017-11-08            2   
2526  777162451382   451382             1 2017-11-09            3   
2314  321355221002   221002             1 2017-11-10            4   
2411  748500164328   164328             3 2017-11-11            5   
2953  135342483454   483454             2 2017-11-12            6   
2507  263886601790   601790             2 2017-11-13            0   
2955  312455541811   541811             7 2017-11-14            1   
2401   28411523002    23002             2 2017-11-15            2   
470     8505276782   276782             1 2017-11-16            3   
2400   20232428519   428519             1 2017-11-17            4   
2635  634506366844   366844             2 2017-11-18            5   
87    570727363200   363200             2 2017-11-19            6   
2858   30773137725    37725             1 2017-11-20            0   

                     article_name     article_id  price  quantity  \
2325  Tennis-Socken im Dreierpack  4002702234727   5.01         1   
5836   Basic Socken im Doppelpack  4002702235090   5.84         2   
2849  Tennis-Socken im Dreierpack  4002702234765   5.52         2   
2512  Tennis-Socken im Dreierpack  4002702234758   6.68         2   
455      Geschenkgutschein 50 EUR   200000000050  50.00         1   
357      Geschenkgutschein 50 EUR   200000000050  50.00         1   
4025   Basic Socken im Doppelpack  4002702235076   5.84         2   
2574  Tennis-Socken im Dreierpack  4002702234758   4.49         2   
2526  Tennis-Socken im Dreierpack  4002702234758   6.68         1   
2314  Tennis-Socken im Dreierpack  4002702234727   6.68         1   
2411  Tennis-Socken im Dreierpack  4002702234734   4.01         2   
2953  Tennis-Socken im Dreierpack  4002702234765   6.68         1   
2507  Tennis-Socken im Dreierpack  4002702234758   4.01         2   
2955  Tennis-Socken im Dreierpack  4002702234765   4.01         1   
2401  Tennis-Socken im Dreierpack  4002702234734   4.01         1   
470      Geschenkgutschein 50 EUR   200000000050  50.00         1   
2400  Tennis-Socken im Dreierpack  4002702234734   4.01         1   
2635  Tennis-Socken im Dreierpack  4002702234758   4.01         1   
87       Geschenkgutschein 10 EUR   200000000010  10.00         1   
2858  Tennis-Socken im Dreierpack  4002702234765   5.25         1   

      total_price  shop_cart_value  shop_cart_item_count  \
2325         5.01            32.69                     5   
5836        11.68           108.55                     5   
2849        11.04            83.18                     7   
2512        13.36            78.38                     4   
455         50.00            50.00                     1   
357         50.00            50.00                     1   
4025        11.68           115.76                     9   
2574         8.98            63.59                     7   
2526         6.68            20.12                     2   
2314         6.68            38.56                     3   
2411         8.02            90.59                    13   
2953         6.68            40.28                     3   
2507         8.02           106.31                     7   
2955         4.01           150.18                     5   
2401         4.01           276.18                    17   
470         50.00            50.00                     1   
2400         4.01           159.23                    12   
2635         4.01           12

## features for first date of purchase and a feature for first or established purchase

In [118]:
#help feature with a edited date
df['year_month'] = df['date'].map(lambda date: 100*date.year + date.month)

In [119]:
#finding the date where users bought for the first time in our timeline
min_purchase = df.groupby("user_id").year_month.min().reset_index()
min_purchase.columns = ["user_id","min_purchase_date"]

#merge
df = pd.merge(df, min_purchase, on="user_id")

In [125]:
#create a feature user_type with first or established depending on time of purchase
df["user_type"] = "first"
df.loc[df["year_month"]>df["min_purchase_date"], "user_type"] = "established"

### feature dictionary after some work

| Feature No. | Feature Name | Description |
| :-: | :-: | :-: |
| 1 | visit_id | Unique ID consisting of the ID of a single shopping cart and the ID of the user |
| 2 | user_id | Unique ID of a user |
| 3 | visit_number | Not entirely clear; either number of visits to the shop or number of total purchases |
| 4 | date | Date the purchase was made |
| 5 | day_of_week | Day when the purchase was made; 0 for monday till 6 for sunday |
| 6 | article_name | Name of the article that was bought |
| 7 | article_id | ID of the article that was bought |
| 8 | price | Price of the article |
| 9 | quantity | Number of the article that were bought |
| 10 | total_price | Total price of the article purchase consisting of price * quantity |
| 11 | shop_cart_value | Value of the whole shopping cart of a single visit_id |
| 12 | shop_cart_item_count | Number of items in the shopping cart of a single visit_id |
| 13 | shop_cart_unique_articles | Number of unique items in the shopping cart of a single visit_id |
| 14 | country | country location of the user |
| 15 | region | region location of the user |
| 16 | city | city location of the user |
| 17 | city_id | geoID of the town location of the user |
| 18 | women_article | women category - 1 yes , 0 no |
| 19 | men_article | men category - 1 yes , 0 no |
| 20 | kids_article | kids category - 1 yes , 0 no |
| 21 | homeandliving_article | home & living category - 1 yes , 0 no |
| 22 | campaign_involved | still not completely clear but after some research it might stand for the cases where users came to the shop via advertisement. When you get to the shop this way the url contains a part for campaign: for example "utm_campaign=de_retargeting_criteo"|
| 23 | category | Category the article falls in ( partly the path in the webshop to the article ) |
| 24 | category_slim | featured category with women, men, kids and home&living |
| 25 | brand | Brand of the article |
| 26 | product_line | Product line of the article |
| 27 | product_group | Product group of the article |
| 28 | product_line_r | not clear what the r stands for; might stand for retail |
| 29 | product_group_r | not clear what the r stands for; might stand for retail |
| 30 | genre_column_placeholder | Sum of article name, category and product lines/groups to better work with |
| 31 | device_category | Category the user device falls in |
| 32 | device_browser | Category the user browser falls in |
| 33 | device_os | Category the user OS falls in |
| 34 | stock_value | company stock price at this day |
| 35 | year_month | feature with year and month as int |
| 36 | min_purchase_date | date of the first purchase in our timeline of this user |
| 37 | user_type | type of the purchase, first or established |

## save dataframe to pickle

In [130]:
df.to_pickle('Data/dataframefull.pickle')